### PR TITLE
Return a better error when a user tries to cast a document that has already been accessed to a value

### PR DESCRIPTION
**Pull request by lemire**
In https://github.com/simdjson/simdjson/issues/2084, @mmatrosov reported an error with the following code...

```cpp
      padded_string json(R"({"foo": "bar"})"sv);
      ondemand::parser parser;
      ondemand::document doc = parser.iterate(json);
      // Next line starts iterating through the document as an object, effectively
      // instantiating the object.
      std::string_view foo = doc["foo"];
      // This next line tries to a value instance representing the object as a value.
      // It is not allowed because the object instance is already in play.
      simdjson::ondemand::value value = doc;
```

Currently, this fails with a somewhat cryptic error "A JSON document made of a scalar (number, Boolean, null or string) is treated as a value" whether we compile in debug or release mode.

Compiling such code in debug mode should now return an OUT_OF_ORDER_ITERATION message which should help debugging.

I have also clarified the documentation.

### CODE REVIEW

File: doc/basics.md
Comment (mmatrosov) [inline doc/basics.md:545]: "arary"

File: include/simdjson/generic/ondemand/document-inl.h
Comment (mmatrosov) [inline include/simdjson/generic/ondemand/document-inl.h:72]: Could you please clarify why this if under #ifdef, while the check `iter.assert_at_root();` below is always done? If the performance is your concern, wouldn't it worth to put both under #ifdef? If not, maybe move both out of it?

File: include/simdjson/generic/ondemand/document.h
Comment (mmatrosov) [inline include/simdjson/generic/ondemand/document.h:282]: "iSCALAR_DOCUMENT_AS_VALUE" - strange case

File: include/simdjson/generic/ondemand/document-inl.h
Comment (mmatrosov) [inline include/simdjson/generic/ondemand/document-inl.h:74]: BTW in our wrapper we always call `rewind()` when casting a document to a value, and it works even if the iteration was started.

File: include/simdjson/generic/ondemand/document.h
Comment (lemire) [reply include/simdjson/generic/ondemand/document.h:282]: Much appreciated.

File: include/simdjson/generic/ondemand/document-inl.h
Comment (lemire) [reply include/simdjson/generic/ondemand/document-inl.h:74]: If you rewind the document, yes, it would work. I will add a test and documentation for that case.

File: include/simdjson/generic/ondemand/document-inl.h
Comment (lemire) [reply include/simdjson/generic/ondemand/document-inl.h:72]: It is the convention throughout the code that  the macro `SIMDJSON_DEVELOPMENT_CHECKS` determines whether we check for OUT_OF_ORDER_ITERATION. The logic behind it is that these errors only occurs when the code that was written while breaking some simdjson::ondemand requirement. They should not occur in released code after these issues were fixed.

The `iter.assert_at_root();` is different. In Debug mode, when `SIMDJSON_DEVELOPMENT_CHECKS` is unset, it may not serve a purpose in this case, but when compiling in release mode, it generates SIMDJSON_ASSUME statements which instruct the compilers to make some assumptions. 

I will add comments to the code.

File: doc/basics.md
Comment (lemire) [reply doc/basics.md:545]: Thank you.

File: include/simdjson/generic/ondemand/document-inl.h
Comment (lemire) [reply include/simdjson/generic/ondemand/document-inl.h:81]: Test added and documentation improved.

File: include/simdjson/generic/ondemand/document-inl.h
Comment (lemire) [reply include/simdjson/generic/ondemand/document-inl.h:75]: Comments added.

**Comment by lemire** [comment]
@mmatrosov  Do you approve?

**Review by mmatrosov — APPROVED** [review]
Yeah, sorry, I did not think you was waiting for my approval

**Comment by lemire** [comment]
@mmatrosov Thanks.

**Comment by lemire** [comment]
This will be part of the next release.

**Pull request merged**
